Delta auto-calibration updates

LVD-AC opened the pull request:
Thanks to @teemuatlut, I was able to recreate

LVD-AC committed:
M665 rework and related issues (all delta's)

- Making M665 compatible with repetier (see
http://reprap.org/wiki/G_code#M665:_Set_delta_configuration)

- M665 B also sets the radius for manual calibration menu

- Converting tower ajustment definitions to arrays - tower angle
corrections compatible with Esher 3D wizzard

- Only tower angles need to be adjustable with M665 and stored to EEPROM
- tower radius and diag rod can be adjusted in the FW only with #define

LVD-AC committed:
Delete 'expert level'

and getting rid off verbose level 3 and configuration_adv settings

LVD-AC committed:
Adding upto 7*7 probe grids

- adding Cn 5, 6 and 7

LVD-AC committed:
Tower angles

Adding '7-point' tower angle correction calibration

LVD-AC commented:
Need some help here, resolving the conflicts

teemuatlut commented:
Rebased and resolved conflicts.

I've done what I can without knowing more about the subject so if you can take a quick look everything is there. I only had to go through the configuration_store.cpp file.

LVD-AC commented:
I was just one bracket short? The rest seems to be fine... Thanks a lot

LVD-AC commented:
Before this is merged I have 2 small changes to add. How do I proceed? I open this PR in my desktop - made the changes but I can't publish: no write access...

teemuatlut commented:
You can make the PR against my repo and I'll merge it then.

LVD-AC commented:
You lost me again, how do I do that? Whatever I try Desktop tells me I do not have permission to do it... I'm done with GitHub, I'm not allowed to do anything whatsoever.

teemuatlut commented:
I've given you push access to my repo for a while. Maybe that'll help.

LVD-AC commented:
No luck, I'm no longer allowed to clone your branch now... (a debug log of 3 pages cryptic error messages). I can no longer make PR either some 'validation failed' error. GitHub doesn't like me and the feeling is very mutual. 

I have only made 2 little changes in marlin_main: https://github.com/LVD-AC/Marlin-LVD/commit/1b3e677711142e3297c7856f076cf3c3774682c6

teemuatlut committed:
Delta probe height bug solved

thinkyhead commented on Marlin/Conditionals_post.h:
```diff
@@ -691,43 +691,20 @@
-    #ifndef DELTA_TOWER_ANGLE_TRIM_2
-      #define DELTA_TOWER_ANGLE_TRIM_2 0.0
+    #ifndef DELTA_TOWER_ANGLE_TRIM
+      #define DELTA_TOWER_ANGLE_TRIM {0.0, 0.0}
```
Three elements here?

thinkyhead commented on Marlin/configuration_store.cpp:
```diff
@@ -92,12 +92,10 @@
+ *  372  M665 B    delta_calibration_radius         (float)
+ *  376  M665 X    delta_tower_angle_trim[A]        (float)
+ *  380  M665 Y    delta_tower_angle_trim[B]        (float)
+ *  ---  M665 Z    delta_tower_angle_trim[C]        (float) is always 0.0
```
If the EEPROM layout changes, the EEPROM version should change too.

LVD-AC replied:
Indeed, I leave that EEPROM version up to you, so it's done properly

thinkyhead replied:
Just increment the version number by 1.

LVD-AC replied:
@thinkyhead  Done

thinkyhead replied:
Not seeing it reflected in this PR yet.

thinkyhead replied:
No need to include `delta_tower_angle_trim[C]` here.
For one thing, it doesn't exist. The array now has only A and B.

LVD-AC replied:
EEPROM version was updated in https://github.com/LVD-AC/Marlin-LVD/commit/46a6dfd33e2380a10a7e6839b8eecab665f19d88

thinkyhead commented on Marlin/example_configurations/delta/generic/Configuration.h:
```diff
@@ -461,28 +461,30 @@
 
+
+
+  // set the radius for the calibration probe points - max 0.8 * DELTA_PRINTABLE_RADIUS if DELTA_AUTO_CALIBRATION enabled
```
Probably don't need these extra blank lines.

LVD-AC replied:
@thinkyhead Done

thinkyhead commented:
Overall looking pretty good. Only small spacing changes required. If needed, I can apply cleanup and squash to a single commit before merging.

LVD-AC commented on Marlin/configuration_store.cpp:
```diff
@@ -92,12 +92,10 @@
+ *  372  M665 B    delta_calibration_radius         (float)
+ *  376  M665 X    delta_tower_angle_trim[A]        (float)
+ *  380  M665 Y    delta_tower_angle_trim[B]        (float)
+ *  ---  M665 Z    delta_tower_angle_trim[C]        (float) is always 0.0
```
Indeed, I leave that up to you, so it's done properly

LVD-AC replied:
Indeed, I leave that EEPROM version up to you, so it's done properly

LVD-AC committed:
Allowing #define DELTA_TOWER_ANGLE_TRIM[C_AXIS]

More consistent with M665 where C-value is allowed as well, then
normalized

LVD-AC commented:
> +      #define DELTA_TOWER_ANGLE_TRIM {0.0, 0.0}
> 
> Three elements here?

The tower angles are normalized by rotating all three (ABC) until C-tower = zero.

Two options here:

- not allowing the user to change C-value with #define and/or M665 
- allowing the input of the C-axis value and normalizing them after input

I made a mix of both approaches : so either disable the 'M665 Z' option or apply commit https://github.com/LVD-AC/Marlin-LVD/commit/065b915502a2fa00624e492a44cb9751830fecfd that allows the input of 3 values in the #define statement and normalizes them internally (makes more sense to me)

@teemuatlut : can you add that commit as well
@thinkyhead : please go ahead, to clean, squash and merge after that

thinkyhead commented:
> please go ahead, to clean, squash and merge

Pretty soon. Busy days.

LVD-AC commented:
Take your time. Thanks for the help.

LVD-AC committed:
z_offset change

Avoids to have to recalibrate the delta_height after a z_offset change

LVD-AC commented:
@teemuatlut : One more commit to add, if you can beat @thinkyhead to it: https://github.com/LVD-AC/Marlin-LVD/commit/4688393e963f716f0e40e2c225624c19ada01a7a

LVD-AC committed:
Cn negative : no tower angle calibration

Giving a negative number of probe points disables the tower angle
correction calibration ('4point' instead of '7point' solution)

EEPROM version updated

LVD-AC commented:
@teemuatlut I added another feature : https://github.com/LVD-AC/Marlin-LVD/commit/46a6dfd33e2380a10a7e6839b8eecab665f19d88

LVD-AC committed:
M666 normalize positive values to <=0

M666

+ tower radians absolute value

LVD-AC commented:
@teemuatlut : Flsun FB group is pretty buzzy beta testing. Dotting some i's: https://github.com/LVD-AC/Marlin-LVD/commit/5aece3363c273cf0252adb8f693b3eb27fb607a9

teemuatlut commented:
I'll update tomorrow morning if you have more than one that still needs adding.

LVD-AC committed:
cleanup ready to squash

LVD-AC commented:
@teemuatlut : cleaning and ready to squash now into 1 commit https://github.com/LVD-AC/Marlin-LVD/commit/87fcc44a0ce01dd13f53c2b76c43553508f793c9

teemuatlut commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -446,53 +446,56 @@
   // Center-to-center distance of the holes in the diagonal push rods.
   #define DELTA_DIAGONAL_ROD 218.0 // mm
-
+/*
```
Probably not supposed to do this. If something is not needed, remove it.

LVD-AC replied:
I did delete that part in the Flsun/autocal/configuration.h but @thinkyhead did put it back commented. Probably left it in order to be able to 'roll back'. So just did the same in the other configs.

thinkyhead replied:
If I did put it back commented, that was an error. I don't recall doing so.

LVD-AC replied:
deleted block comments

LVD-AC committed:
Solved "The Travis CI build failed "

LVD-AC commented:
@teemuatlut Solved the 'The Travis CI build failed' error : https://github.com/LVD-AC/Marlin-LVD/commit/2c0db9d840d537e3597a10de60cbe525f52f1082

thinkyhead commented on Marlin/example_configurations/delta/kossel_xl/Configuration.h:
```diff
@@ -443,7 +443,7 @@
   // Center-to-center distance of the holes in the diagonal push rods.
   #define DELTA_DIAGONAL_ROD 317.3 + 2.5 // mm
-
+/*
```
Don't mess with other configurations in this PR, unless related to the feature being modified. Certainly don't leave a block comment here like this.

thinkyhead commented on Marlin/example_configurations/delta/kossel_xl/Configuration.h:
```diff
@@ -452,42 +452,42 @@
   #define DELTA_HOME_TO_SAFE_ZONE
 
-  //#define DELTA_ENDSTOP_ADJ { 0, 0, 0 }
+  #define DELTA_ENDSTOP_ADJ { -0.00, -0.00, -0.00 } // get these from auto calibrate
```
No.

LVD-AC replied:
used 0's everywhere instead

thinkyhead commented on Marlin/example_configurations/delta/kossel_pro/Configuration.h:
```diff
@@ -434,42 +434,42 @@
   #define DELTA_HOME_TO_SAFE_ZONE
 
-  //#define DELTA_ENDSTOP_ADJ { 0, 0, 0 }
+  #define DELTA_ENDSTOP_ADJ { -0.00, -0.00, -0.00 } // get these from auto calibrate
```
Please don't use "`-0.00`".

thinkyhead commented on Marlin/example_configurations/delta/kossel_pro/Configuration.h:
```diff
@@ -434,42 +434,42 @@
   // Horizontal offset of the universal joints on the carriages.
   #define DELTA_CARRIAGE_OFFSET 30.0 // mm
-
+*/
```
Don't leave this block-comment in the configuration.

thinkyhead commented on Marlin/example_configurations/delta/kossel_mini/Configuration_adv.h:
```diff
@@ -1,4 +1,4 @@
-/**
+﻿/**
```
Why do the first lines keep changing? Because of your editor.
Please save as "UTF-8" format, not "UTF-8 with BOM."

LVD-AC replied:
That must be an Arduino thing

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -864,11 +864,11 @@
 #define INVERT_Z_DIR true
 
-// Enable this option for Toshiba stepper drivers
+// Enable this option for Toshiba steppers
```
They are Toshiba **_stepper drivers_** like the TB6600.

LVD-AC replied:
don't remember changing that??? changed it back

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -728,7 +728,7 @@
   #define Z_PROBE_ALLEN_KEY_DEPLOY_2_Y DELTA_PRINTABLE_RADIUS
   #define Z_PROBE_ALLEN_KEY_DEPLOY_2_Z 100.0
-  #define Z_PROBE_ALLEN_KEY_DEPLOY_2_FEEDRATE (XY_PROBE_SPEED)/10
+  #define Z_PROBE_ALLEN_KEY_DEPLOY_2_FEEDRATE (XY_PROBE_SPEED/10)
```
Wrong. The parentheses belong around `XY_PROBE_SPEED` to make it a unit for division. (Set `XY_PROBE_SPEED` to "3000 + 1000" with no parentheses and you'll see why this is done.)

LVD-AC replied:
Fixed

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -704,7 +704,7 @@
 
 // Speed for the "accurate" probe of each point
-#define Z_PROBE_SPEED_SLOW (Z_PROBE_SPEED_FAST / 4)
+#define Z_PROBE_SPEED_SLOW (Z_PROBE_SPEED_FAST / 6)
```
`Z_PROBE_SPEED_FAST` should rightfully also have parentheses around it.

LVD-AC replied:
Fixed

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -667,9 +670,6 @@
 
-// A probe that is deployed and stowed with a solenoid pin (SOL1_PIN)
-//#define SOLENOID_PROBE
-
```
Don't remove this setting, please.

LVD-AC replied:
don't remember deleting that??? put it back

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -601,7 +604,7 @@
 #define DEFAULT_YJERK                 DEFAULT_XJERK
 #define DEFAULT_ZJERK                 DEFAULT_YJERK // Must be same as XY for delta
-#define DEFAULT_EJERK                  5.0
+#define DEFAULT_EJERK                 5.0
```
The extra space is to align the decimal point with "20.0" above.

LVD-AC replied:
OK

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -446,53 +446,56 @@
   #define DELTA_HOME_TO_SAFE_ZONE
 
-  #define DELTA_ENDSTOP_ADJ { -0.05, -0.00, -0.02 } // get these from auto calibrate
+  #define DELTA_ENDSTOP_ADJ { -0.00, -0.00, -0.00 } // get these from auto calibrate
```
Don't use "-0.00" please.

thinkyhead commented on Marlin/example_configurations/delta/FLSUN/auto_calibrate/Configuration.h:
```diff
@@ -1,4 +1,4 @@
-/**
+﻿/**
```
"UTF-8" … not "UTF-8 with BOM."

thinkyhead commented on Marlin/configuration_store.cpp:
```diff
@@ -412,8 +410,8 @@ void MarlinSettings::postprocess() {
       EEPROM_WRITE(delta_segments_per_second); // 1 float
-      EEPROM_WRITE(delta_diagonal_rod_trim);   // 3 floats
-      EEPROM_WRITE(delta_tower_angle_trim);    // 3 floats
+      EEPROM_WRITE(delta_calibration_radius);  // 1 floats
```
"1 float"

LVD-AC replied:
fixed

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5018,47 +5018,32 @@ inline void gcode_G28() {
+      if (!WITHIN(verbose_level, 0, 2)) verbose_level = 1;
 
-      float zero_std_dev = verbose_level ? 999.0 : 0.0; // 0.0 in dry-run mode : forced end
+      float zero_std_dev = verbose_level ? 999.0 : 0.0;                                          // 0.0 in dry-run mode : forced end
```
No need for extra spaces here.

LVD-AC replied:
removed all extra spaces

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5018,47 +5018,32 @@ inline void gcode_G28() {
+            zh_old = home_offset[Z_AXIS],
+            alpha_old = delta_tower_angle_trim[A_AXIS],
+            beta_old = delta_tower_angle_trim[B_AXIS];
       COPY(e_old,endstop_adj);
```
```cpp
const float e_old[XYZ] = {
              endstop_adj[A_AXIS],
              endstop_adj[B_AXIS],
              endstop_adj[C_AXIS]
            },
            dr_old = delta_radius,
            zh_old = home_offset[Z_AXIS],
            alpha_old = delta_tower_angle_trim[A_AXIS],
            beta_old = delta_tower_angle_trim[B_AXIS];
```

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5018,47 +5018,32 @@ inline void gcode_G28() {
+            zh_old = home_offset[Z_AXIS],
+            alpha_old = delta_tower_angle_trim[A_AXIS],
+            beta_old = delta_tower_angle_trim[B_AXIS];
       COPY(e_old,endstop_adj);
```
Add a space:
```cpp
COPY(e_old, endstop_adj);
```

LVD-AC replied:
deleted (double assignment)

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5018,47 +5018,32 @@ inline void gcode_G28() {
       LCD_MESSAGEPGM("Checking... AC");
 
-      SERIAL_PROTOCOLPAIR("Height:", DELTA_HEIGHT + home_offset[Z_AXIS]);
+      SERIAL_PROTOCOLPAIR(".Height:", DELTA_HEIGHT + home_offset[Z_AXIS]);
```
This thing with sticking the dot character in front of log lines: don't do that.
Nothing else in Marlin does it, and it sets a bad precedent.

LVD-AC replied:
Indeed everything from Marlin is wrongly aligned in Repetier- Host as well as in Pronterface  (eg G29 output the column numbers do not correspond to the table below) because of those leading spaces being suppressed (very enjoying feature of Marlin not taking this into account). The output from the iteration routine is already difficult to get, so if everything is all over the place it is pure mayhem. And I do not think I'm the only one using those apps.

Hopefully this will set a good precedent and ALL the Marlin output would become more readable.

Roxy-3D replied:
> Indeed everything from Marlin is wrongly aligned in Repetier- Host as well as in Pronterface (eg G29 output the column numbers do not correspond to the table below) because of those leading spaces being suppressed (very enjoying feature of Marlin not taking this into account).

Both Repetier and PronterFace use proportional fonts.  So spaces, dashes, decimal points all take different amounts of room.     Different laptops use different versions of the fonts.  It is impossible to get those items to line up nicely in all cases when you have to display things in a proportional font.

LVD-AC replied:
Repetier used a fixed space font; it remains only a Ponterface problem

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5084,91 +5079,96 @@ inline void gcode_G28() {
           N++;
         }
-        zero_std_dev = round(sqrt(S2 / N) * 1000.0) / 1000.0 + 0.00001; // deviation from zero plane
+        zero_std_dev = round(sqrt(S2 / N) * 1000.0) / 1000.0 + 0.00001;                          // deviation from zero plane
```
No need for extra spaces here.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5276,24 +5254,24 @@ inline void gcode_G28() {
+            LCD_MESSAGEPGM("Calibration OK");
           }
-          else {                                // !end iterations
+          else {                                                                                 // !end iterations
```
No need for extra spaces here. If your editor is causing this stuff, please change your editor settings.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5307,20 +5285,32 @@ inline void gcode_G28() {
+          }
           if (zero_std_dev >= test_precision)
-            SERIAL_PROTOCOLLNPGM("Save with M500");
+            SERIAL_PROTOCOLLNPGM("save with M500 and/or copy to configuration.h");
```
The filename is capitalized, and the sentence should also be capitalized.
"Save with M500 and/or copy to Configuration.h"

LVD-AC replied:
Fixed

thinkyhead commented on Marlin/Conditionals_post.h:
```diff
@@ -691,43 +691,20 @@
-    #ifndef DELTA_TOWER_ANGLE_TRIM_2
-      #define DELTA_TOWER_ANGLE_TRIM_2 0.0
+    #ifndef DELTA_TOWER_ANGLE_TRIM
+      #define DELTA_TOWER_ANGLE_TRIM {0.0, 0.0, 0.0}
```
If this setting is only going to have 2 items, with the third item always being zero, then make this 2 items long, not 3.

LVD-AC replied:
The user may enter a z_value to rotate the z-tower, but it will be normalized to zero in the calculations; so I opted for: user interface = 3 values; internal calculations 2 values only.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5084,91 +5079,96 @@ inline void gcode_G28() {
         int16_t center_points = 0;
 
-        if (probe_points != 3) {
+        if (abs(probe_points) != 3 &&  abs(probe_points != 6)) {                                 // probe centre
```
Since this uses `abs(probe_points)` continually, please create `const uint8_t abs_probe_points` and use that instead. Also a `const bool alt = (probe_points < 0)` may be useful. i.e.,
```cpp
const int8_t cv = code_seen('C') ? code_value_int() : DELTA_CALIBRATION_DEFAULT_POINTS,
             c_value = (WITHIN(cv, -7, -1) || WITHIN(cv, 1, 7)) ? cv : DELTA_CALIBRATION_DEFAULT_POINTS;

const uint8_t probe_points = abs(c_value);
```

thinkyhead replied:
Reducing redundancies can help avoid glitches like…
```
abs(probe_points != 6)
```

LVD-AC replied:
created abs_probe_points; fixed

thinkyhead commented:
I would rather that contributors manage their own cleanups, but time is short with the release due Sunday, so I'm being magnanimous. But in future, please don't expect other volunteers to do the cleanup work on your behalf. Git can be mastered in a weekend.

thinkyhead commented:
Please add documentation for `G33` to the RepRap wiki at your earliest convenience.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4982,32 +4988,26 @@ inline void gcode_G28() {
+     *          n=-2 same but opposite towers
+     *          n=3  probes all points: center, towers and opposite towers / sets all
+     *          n>3  probes all points multiple times and averages
+     *          n<=3 same but tower angle calibration disabled
```
This can't possible be correct. `C0` through `C3` are described above as having specific effects. Now this is saying that `C0` to `C3` "probe all points multiple times and average with tower angle calibration disabled." Furthermore, the comment above refers to "n>=3" and then here you have "n>3." So which is it?

I will do my best to make sense of the code and edit this comment for clarity.

thinkyhead replied:
I'm not fond of this overloading of the `C` parameter with magic numbers. I would prefer that there be more parameters with specific meaning rather than assigning special meaning to `C0`, `C1`, and `C2`. Meanwhile, most other GCodes that need a "number of points" input use the parameter `P`. I am inclined to re-write the entire `gcode_G33` function to eliminate these magic numbers and follow established patterns.

LVD-AC replied:
Like this?

    /**
     * G33 - Delta '4-7-point' auto calibration (Requires DELTA)
     * 
     * Usage:
     *   G33 <Pn> <A> <O> <T> <Vn>
     *   
     *     Pn = n=-7 -> +7 : n*n probe points
     *          calibrates height ('1 point'), endstops, and delta radius ('4 points') 
     *          and tower angles with ('7+ points')
     *          n=1  probes center / sets height only
     *          n=2  probes center and towers / sets height, end-stops and delta radius
     *          n=3  probes all points: center, towers and opposite towers / sets all
     *          n>3  probes all points multiple times and averages
     *     A  = abort 1 point delta height calibration after 1 probe
     *     O  = use oposite tower points instead of tower points with 4 point calibration
     *     T  = do not calibrate tower angles with 7+ point calibration
     *     Vn = verbose level (n=0-2 default 1)
     *          n=0 dry-run mode: no calibration
     *          n=1 settings 
     *          n=2 setting + probe results 
     */

Is already adapted like this

thinkyhead replied:
I propose to replace `C1` / `C-1` with…
```cpp
    /**
     *   Z      Probe the center to set the Z height
     *   Z1     Same but 1 iteration only
     */
```

LVD-AC replied:
It's not even one iteration; just one quick probe and done ;) But why make an exception for 1-point calibration: P1 and P1 A, then P2 and P2 O, P3, P4,... P7 and P3 T, P4 T... P7 T there's more logic to it. And people who do not remember all those parameters just use P1 to P7 the A, O and T parameters are more 'expert level' anyhow.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4982,32 +4988,26 @@ inline void gcode_G28() {
-     *    2: Setting + probe results
-     *    3: Expert mode: setting + iteration factors (see Configuration_adv.h)
-     *       This prematurely stops the iteration process when factors are found
+     * G33 - Delta '4-7-point' auto calibration (Requires DELTA)
```
Ahem…
```cpp
#if ENABLED(DELTA_AUTO_CALIBRATION)
  . . .
     * G33 - Delta '4-7-point' auto calibration (Requires DELTA)
```

thinkyhead committed:
DELTA_TOWER_ANGLE_TRIM only needs AB

thinkyhead committed:
Starting on G33 rewrite

Roxy-3D commented:
My suggestion would be to leave this until next week when v1.1.0 is tagged...

> Git can be mastered in a weekend.

Oh???   I guess I'm a very slow learner.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5084,91 +5079,96 @@ inline void gcode_G28() {
-          if (probe_points == 4) step_axis = 2;
+        step_axis = (abs(probe_points) == 2) ? 4 : (abs(probe_points) == 4 || abs(probe_points) > 5) ? 1 : 2;
+        float start_circles = (abs(probe_points) > 6) ? -1.5 : (abs(probe_points) > 4) ? -1 : 0, // one or multi radius points
+              end_circles = (abs(probe_points) > 6) ? 1.5 : (abs(probe_points) > 4) ? 1 : 0;     // one or multi radius points
```
So…
```cpp
end_circles = -start_circles;
```

LVD-AC replied:
Fixed

thinkyhead committed:
Add a function to output spaces to serial

thinkyhead committed:
Initial patches to G33

teemuatlut commented:
@LVD-AC I think that any work on deltas for Marlin is highly valuable but as the new release is just around the corner, maybe it would be best if after the fixes and cleanup you would create a new PR that you can manage yourself. This would be easier for you and especially me =)

LVD-AC commented:
@teemuatlut , @thinkyhead , @Roxy-3D here's a commit that deals with the remarks https://github.com/LVD-AC/Marlin-LVD/commit/948e9c6aa495953d1d9e11864fc7ac8994ffad33 ; squash and merge as and when you see fit.

thinkyhead commented:
> My suggestion would be to leave this until next week

I would rather get this fixed than have a `G33` in the release that's only going to be redefined days later.

> Oh??? I guess I'm a very slow learner.

It can (as I can testify). Lots of online tutorials, even interactive ones. Just have to put in the effort.

thinkyhead commented:
Thanks @LVD-AC — I did start a new PR, and I will work on integrating the applicable changes from your last commit. #6492

LVD-AC commented:
> Please add documentation for  G33  to the RepRap wiki at your earliest convenience.

I've asked for a login...

LVD-AC commented:
@thinkyhead @teemuatlut @Roxy-3D 
> Thanks @LVD-AC — I did start a new PR, and I will work on integrating the applicable changes from your last commit. #6492 

Yet another version to work on (and I can't get into my desktop)??? I had a look at it and made similar changes (easy done) to this PR in https://github.com/LVD-AC/Marlin-LVD/commit/00a4cee60474d13ec71cff09d39d0f214974206c . Wouldn't it be better to reopen and continue here, with the finish line in sight instead of starting all over again?

thinkyhead commented:
> Wouldn't it be better to reopen and continue here

Sure. You just have to rebase your working branch on the latest `RCBugFix` first. Then I will be able submit the additional patches to your branch.

LVD-AC commented:
@thinkyhead Give me a break!

I'm an architect, I make buildings, I'm not a software engineer and since I didn't have a weekend yet to get me masters degree from Harvard in IT-technology, I have still no clue on how to do that.

@teemuatlut is able to add additional commits, but I have to ask him since whatever I do Git is giving me fatal errors.

In the mean time I did follow some guide about Git Hub and the result is I no longer can make commits to this branch either... cryptic errors from the 1st commands I try to enter in Git Shell.

I can't seem to get past this whatever I try:

```
~\Documents\GitHub\Marlin-LVD\Marlin [delta_tower_calibration ≡]> Git fetch upstream
remote: Repository not found.
fatal: repository 'https://Github.com/MarlinFirmware/Marlin.Git/' not found
```

Out of frustration I deleted everything I could find on my computer in order to start over from frech... but it's always referring to what is stored on the Git server and it's getting only worse... I'm gonna reset my computer to a restore point before Git, and if I can't find a way to restore my Git account back to virgin state I'll delete it and make a new log-in.

thinkyhead commented:
> I'm not a software engineer 

Ok, so you see my point. I made the correct choice to start over with a fresh branch.

thinkyhead commented:
> I can't seem to get past this whatever I try

You misspelled the path to the repository.

```
git remote remove upstream
git remote add upstream git@github.com:MarlinFirmware/Marlin.git
git fetch upstream
```

LVD-AC commented:
Your on your own then, since the only way I can view this is with notepad and that is undoable. See ya with a new account in a bit...

thinkyhead commented:
You don't need to create a new account just to fix a messed-up branch.

LVD-AC commented:
> You misspelled the path to the repository

I copy-pasted it from the marlinfw.org guide

thinkyhead commented:
> I copy-pasted it from the marlinfw.org guide

Sorry, it would not be like that if I had had a chance to review the documentation before it was posted.

LVD-AC commented:
Ok I managed to re-clone marlin in my desktop - what do I do next to get this PR in it?

thinkyhead commented:
This PR is under teemuatlut's account, so you need to fetch the branch from there.
```
git remote add teemuatlut git@github.com:teemuatlut/Marlin.git
git fetch teemuatlut
git checkout teemuatlut/LVD-Delta -b lvd_delta_work
git push --set-upstream origin lvd_delta_work
```

thinkyhead commented:
If that works, then I will try submitting my patches to your branch. If it needs a rebase, I will guide you in that process.

thinkyhead commented:
Once you have that branch set up, then you're ready to make commits that can be submitted as PRs to teemautlut's fork. But let me try submitting my patches to his fork first, and then see what more you want to patch in `G33` after that, as I've done a lot of preliminary cleanup, but left it functionally the same.

LVD-AC commented:
So far so good I good:

This branch in my desktop
fixed that upstream error

So how do I "You just have to rebase your working branch on the latest  RCBugFix  first"

And how do I get my 2 last commits into my working branch?

thinkyhead commented:
> So how do I "You just have to rebase your working branch on the latest RCBugFix first"

At this point, we are re-opening this PR. So after my PR at teemuatlut/Marlin#2 is merged, you can pull down those changes to your branch, and then we can continue from there.

Once my changes have been merged into this PR, then you will need to do the following:
```
git fetch teemuatlut
git checkout lvd_delta_work
git merge teemuatlut/LVD-Delta
```
…and then your branch will be ready for additional changes.

LVD-AC commented:
did: 
git fetch teemuatlut
git checkout lvd_delta_work
git merge teemuatlut/LVD-Delta

But nothing comes up in desktop??? Do I need to close that process with something like "Git push –f"?

Tried it but still nothing in the desktop?

**OK got it I have to go trough all the files to see the changes made**

LVD-AC committed:
Proposed changes

This is what I did yesterday:

- basicly gave the tests more comprehensive names; put all the
declarations at the top; got rid of the magic negative C-value (renamed
to P + A, O, T)

- "cos(RADIANS(180 + 30 * axis)) * (1 + circles * 0.1 * ((zig_zag) ? 1 :
-1)) * delta_calibration_radius" compiles wrong is zig_zag statement is
without brackets

- DELTA_TOWER_ANGLE_TRIM reset to 3 values (the calcs use the 3th value
to normalize will not compile otherwise)

-Wrote 3 dummies to keep EEPROM lenght the same

-Reset the configs to the 'original' with autocal + menu disabled (but
can be enabled of course)

LVD-AC commented:
I made 2 commits, synced it, squashed it to one commit https://github.com/LVD-AC/Marlin/commit/49d09f9c21e72550891fe1126ff5adc32a4b9980 but when I try to make a pull request it says the usual error:

"Request failed - validation error"

@thinkyhead what now???

Didn't work in desktop, but did work online ???? Strange : https://github.com/teemuatlut/Marlin/pull/3

LVD-AC commented:
I do not know what happend in the last BugFix but my printer almost went through the roof just homing it. The speed went up by at least a factor 10. All 3 carriages cracked when trying to send the end-stops in outer space. This is undoable and this is nothing I changed. My non Git copy were I put all the same changes as in this PR (except all the changes of the last days that had nothing to do with me in RCBugFix branche) is still performing normally. I just flag this up since you want to release tomorrow. And there is serious problem there.

LVD-AC commented:
now I'm done @teemuatlut @thinkyhead  thanks for all the help

LVD-AC committed:
leaner code for probe routine

LVD-AC committed:
little oops

thinkyhead commented:
> Didn't work in desktop, but did work online

Best to avoid using Github Desktop except for the simplest things like making commits, comparing changes, and making simple PRs to targets that it knows about. It forces the use of "`git merge`" when most of the time "`git rebase`" is much better.

thinkyhead commented:
So I'm curious, what happens when you use the command like `git rebase -i upstream/RCBugFix` on your system? Does it open up an editor so that you can fix up commits, or does it just bark?

LVD-AC commented:
It opens notepad to squash (if no conflicts or errors were encountered) but for fixing conflicts it loads a file in desktop, I'm using Arduino to edit them. (I have been playing a bit since @thinkyhead pointed me in the right direction by fixing that wrong upstream that blocked me from everything). (y)

LVD-AC commented:
That speed problem I had/have, seems to be solved in the RCBugFix branche, so I guess that will be OK once this is merged.

thinkyhead commented:
> It opens notepad to squash … I'm using Arduino to edit them

Ouch. I'd recommend Sublime Text 3. It's great. With the DevIoT plugin you can build and upload Marlin from within the editor. It also has github integration and other excellent features. Makes working on the code so much easier.

I also hear NotePad++ is good for doing diffs — with the right plugin.

LVD-AC commented:
So we made the deadline?

thinkyhead commented:
Looks that way! And we have an extra 18 hours for last-minute patches….

LVD-AC commented:
Like #6523 : change that EEPROM version @bgort spotted; managed to make that PR on my own - pretty proud it went without git spitting errors all around...

thinkyhead commented:
> pretty proud it went without git spitting errors all around

Haha! The more I mess with git the better my luck seems to become.

LVD-AC commented:
> Please add documentation for  G33  to the RepRap wiki at your earliest convenience.

done : http://reprap.org/wiki/G-code#G33:_Delta_Auto_Calibration_.28Marlin_1.1.0.29

thinkyhead commented:
Good to have documentation!

But now as I look at the RepRap wiki, it seems like `G33` should be changed to `G34` (or something else) and that our UBL `G26` should be changed to `G33`, because that would match up more closely with the definition of `G33` for the two other firmwares that implement it. And, `G33` does actually look like it would be a good thing for Marlin to have (instead of overloading `G29` as we now do).

thinkyhead commented:
Actually, meh, `G26` is its own thing. Still, we _could_ theoretically support `G33` for compatibility with those other firmwares… Why did we select `G33` for auto-calibration?

Roxy-3D commented:
We can move G26 to G33.   But I do think 'Mesh Validation Pattern' is going to be important enough to have its own number.   We can move it up to G123 or some other easy to remember number.

thinkyhead commented:
`G26` is fine.

The time is nigh to document your `G26` and the UBL version of `G29` on the [RepRap wiki GCode page](http://reprap.org/wiki/GCode), so have at it! We're supposed to document them before we publish, to "claim" the codes.

Roxy-3D commented:
> The time is nigh to document your G26 and the UBL version of G29 on the RepRap wiki GCode page, 

I tried to do that.   It won't give me a log in to make changes.   They shut down the credential process and only people that originally registered can make any changes to the Wiki.

thinkyhead commented:
Twats.

Well, here's the page source code. Edit it as you wish, and then post the updated version here and I will publish it for you.

[GCode.txt](https://github.com/MarlinFirmware/Marlin/files/975021/GCode.txt)

Roxy-3D commented:
Scott,   We operate differently...   I tried several months ago to edit their Wiki and keep the GCodes up to date and try to help everybody out.   I went through the whole registration process and tried and tried to get things updated.   THEY DIDN'T EVEN BOTHER TO HAVE PROPER ERROR MESSAGES SAYING WHAT THE PROBLEM WAS.

Screw them.   Marlin **IS** the dominate RepRap firmware.   If they can't let an Owner of Marlin update their Wiki...  PISS ON THEM!

Maybe somebody that knows the self-important keepers of that Wiki will forward this post to them.

LVD-AC commented:
I just did send an e-mail asking for a login, 2 days later I got a message back with my password... ???

And why G33, because a year ago when I started developing this it was available...

Roxy-3D commented:
> I just did send an e-mail asking for a login, 2 days later I got a message back with my password... ???

Well... They have a bunch of attempts from me to try to get credentials...   Let's see if I get an email.   

And for shit's sake...   It's a Wiki...  If somebody is willing to spend time helping them improve their content... They want to put up barriers?   Guess What???   I can assign any open GCode or MCode that I want.   If they don't want me to help them keep the list up to date...  As I add new functions...  There is no reason to even consult their list.   It is obviously out of date and not accurate.

LVD-AC commented:
It's a bit of a search but here's the procedure: http://forums.reprap.org/read.php?1,641102

Roxy-3D commented:
Ahhh...   Did you even read that thread?   Guess what???   

I have a post in that thread.  Please check out:  http://forums.reprap.org/read.php?1,641102,686449#msg-686449

Obviously...  none of the Powers To Be even read that thread....

 I AM NOT GOING THROUGH THE PROCESS AGAIN.   

If they want me to have credentials...  They can choose to activate one of my prior attempts to get credentials.
